pandas/io/parsers.py

```diff
@@ -1425,12 +1425,14 @@ def __init__(self, src, **kwds):
         self.orig_names = self.names[:]
 
         if self.usecols:
-            if len(self.names) > len(self.usecols):
-                self.names = [n for i, n in enumerate(self.names)
-                              if (i in self.usecols or n in self.usecols)]
-
-            if len(self.names) < len(self.usecols):
-                raise ValueError("Usecols do not match names.")
+            if self._reader.file_header is not None:
+                h = self._reader.file_header[0]
+                usecol_len = len(set(self.usecols) - set(h))
+                usecoli_len = len(set(self.usecols) - set(range(0, len(h))))
+                if usecol_len > 0 and usecoli_len > 0:
+                    raise ValueError("Usecols do not match names.")
```

Comment:
Those code block seems a bit convoluted. Can this be simplified? 
Is there a reason you need `_reader.file_header`. Aren't those already in `names?`

Possibly this would be easier if this check is performed inside `_filter_usecols`. 
Then we could check at the same time for out of bounds integer usecols (eg `usecols=[0,2]` if there are only 2 columns)